src/cpp/src/visual_language/phi3_vision/classes.cpp

```diff
@@ -528,6 +528,196 @@ ov::Tensor hd_feature_transform(const EncodedImage& image_features, InferRequest
+    using namespace ov::op;
+
+    // Get current input batch size (num_crops)
+    auto shape_of = std::make_shared<v3::ShapeOf>(input_nchw);
+    auto axis_0 = v0::Constant::create(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{0});
+    auto axis_0_scalar = v0::Constant::create(ov::element::i64, ov::Shape{}, std::vector<int64_t>{0}); // Axis for Gather
+    auto num_crops = std::make_shared<v8::Gather>(shape_of, axis_0, axis_0_scalar);
+
+    // Calculate required padding amount: padding_needed = max(0, max_crops - num_crops)
+    auto diff = std::make_shared<v1::Subtract>(max_crops_param, num_crops);
+    auto zero = v0::Constant::create(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{0});
+    auto padding_needed = std::make_shared<v1::Maximum>(diff, zero);
+
+    // Configure Pad operation arguments (pads_end)
+    auto zero_3 = v0::Constant::create(ov::element::i64, ov::Shape{3}, std::vector<int64_t>{0, 0, 0}); // Zeros for C, H, W dimensions
+    auto zero_4 = v0::Constant::create(ov::element::i64, ov::Shape{4}, std::vector<int64_t>{0, 0, 0, 0}); // pads_begin
+    auto pads_end = std::make_shared<v0::Concat>(ov::OutputVector{padding_needed, zero_3}, 0);
+
+    // Execute Pad operation (Constant mode, fill with 0)
+    auto pad_value = v0::Constant::create(ov::element::f32, ov::Shape{}, std::vector<float>{0.0f});
+    auto padded = std::make_shared<v1::Pad>(
+        input_nchw,
+        zero_4,      // pads_begin
+        pads_end,    // pads_end
+        pad_value,   // pad_value
+        ov::op::PadMode::CONSTANT
+    );
+
+    return padded;
+}
+
+std::shared_ptr<ov::Model> patch_image_preprocess_into_vision_encoder_model(
+    const std::shared_ptr<ov::Model>& vision_encoder_model,
+    const ProcessorConfig& config) {
+
+    using namespace ov;
+    using namespace ov::op;
+
+    // Input: HD transformed image in NHWC format (uint8)
+    // Shape: {1, -1, -1, 3} => {batch=1, height=dynamic, width=dynamic, channels=3 (RGB)}
+    auto hd_image = std::make_shared<v0::Parameter>(element::u8, PartialShape{1, -1, -1, 3});
+    // Target size for global image resize [height, width]
+    auto global_target_size = std::make_shared<v0::Parameter>(element::i64, PartialShape{2});
+    // Max crops parameter for dynamic padding
+    auto max_crops = std::make_shared<v0::Parameter>(element::i64, PartialShape{});
+
+    // Process global image (resize + normalize + channels_first)
+    auto global_resized = create_bicubic_resize(hd_image, global_target_size);
+    auto global_normalized = create_mean_scale(global_resized, config);
+    auto global_processed = create_channels_first(global_normalized);
+
+    // Process HD image (normalize + channels_first + slice)
+    auto hd_normalized = create_mean_scale(hd_image, config);
+    auto hd_processed = create_channels_first(hd_normalized);
+    auto hd_sliced = create_slice_image(hd_processed);
+
+    // Concatenate global and HD results on GPU
```

Comment:
Applied not only on GPU
```suggestion
    // Concatenate global and HD results
```